store/auth_u1.go

```diff
@@ -0,0 +1,286 @@
+ *
+ */
+
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"gopkg.in/macaroon.v1"
+
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/snapdenv"
+)
+
+var (
+	developerAPIBase = storeDeveloperURL()
+	// macaroonACLAPI points to Developer API endpoint to get an ACL macaroon
+	MacaroonACLAPI   = developerAPIBase + "dev/api/acl/"
+	ubuntuoneAPIBase = authURL()
+	// UbuntuoneLocation is the Ubuntuone location as defined in the store macaroon
+	UbuntuoneLocation = authLocation()
+	// UbuntuoneDischargeAPI points to SSO endpoint to discharge a macaroon
+	UbuntuoneDischargeAPI = ubuntuoneAPIBase + "/tokens/discharge"
+	// UbuntuoneRefreshDischargeAPI points to SSO endpoint to refresh a discharge macaroon
+	UbuntuoneRefreshDischargeAPI = ubuntuoneAPIBase + "/tokens/refresh"
+)
+
+// UserAuthorizer authorizes requests using user credentials managed via
+// the DeviceAndAuthContext.
+type UserAuthorizer struct{}
+
+func (a UserAuthorizer) Authorize(r *http.Request, _ DeviceAndAuthContext, user *auth.UserState, _ *AuthorizeOptions) error {
+	// only set user authentication if user logged in to the store
+	if !user.HasStoreAuth() {
+		return nil
+	}
+
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, `Macaroon root="%s"`, user.StoreMacaroon)
+
+	// deserialize root macaroon (we need its signature to do the discharge binding)
+	root, err := auth.MacaroonDeserialize(user.StoreMacaroon)
+	if err != nil {
+		return fmt.Errorf("cannot deserialize root macaroon: %v", err)
+	}
+
+	for _, d := range user.StoreDischarges {
+		// prepare discharge for request
+		discharge, err := auth.MacaroonDeserialize(d)
+		if err != nil {
+			return fmt.Errorf("cannot deserialize discharge macaroon: %v", err)
+		}
+		discharge.Bind(root.Signature())
+
+		serializedDischarge, err := auth.MacaroonSerialize(discharge)
+		if err != nil {
+			return fmt.Errorf("cannot re-serialize discharge macaroon: %v", err)
+		}
+		fmt.Fprintf(&buf, `, discharge="%s"`, serializedDischarge)
+	}
+	r.Header.Set("Authorization", buf.String())
+	return nil
+}
+
+func (a UserAuthorizer) RefreshAuth(need AuthRefreshNeed, dauthCtx DeviceAndAuthContext, user *auth.UserState, client *http.Client) error {
+	if user == nil || !need.User {
+		return nil
+	}
+	// refresh user
+	return a.RefreshUser(user, dauthCtx, client)
+}
+
+type UserAuthUpdater interface {
+	UpdateUserAuth(user *auth.UserState, discharges []string) (actual *auth.UserState, err error)
+}
+
+// RefreshUser will refresh user discharge macaroon and update state via the UserAuthUpdater.
+func (a UserAuthorizer) RefreshUser(user *auth.UserState, upd UserAuthUpdater, client *http.Client) error {
+	if upd == nil {
+		return fmt.Errorf("user credentials need to be refreshed but update in place only supported in snapd")
+	}
+	newDischarges, err := refreshDischarges(client, user)
+	if err != nil {
+		return err
+	}
+
+	curUser, err := upd.UpdateUserAuth(user, newDischarges)
```

Comment:
nitpick: if `cur` stands for `current`, the most common abbreviation is `curr`; but if this stands for something else, then it's not clear.

Reply:
I always abbreviated current with cur, not curr, go figure, anyway this also just moved code